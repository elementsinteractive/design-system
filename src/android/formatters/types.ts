export type FormatterConfig = {
  template: string
  header: string
  packageName: string
}
