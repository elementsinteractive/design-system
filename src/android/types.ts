export type FormatterConfig = {
  template: string
  header: string
  packageName: string
}

export type ThemeFileOptions = {
  destination: string
  packageName: string
}

export type AndroidTheme = {
  typography: ThemeFileOptions

  colors: ThemeFileOptions

  spacings: ThemeFileOptions
}

export type AndroidJsonOptions = {
  theme: AndroidTheme
}
