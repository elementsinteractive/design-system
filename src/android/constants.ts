import { join } from 'path'
import { AndroidTheme } from './types'

export const TEMPLATES = (theme: AndroidTheme) => ({
  typographyTemplate: {
    source: join(__dirname, './templates/typographies.ejs'),
    destination: `${theme.typography}/Type.kt`,
    filter: 'isTypography',
    formatter: 'android/compose/typography',
    packageName: theme.typography.packageName,
  },
  spacingsTemplate: {
    source: join(__dirname, './templates/spacings.ejs'),
    destination: `${theme.spacings}/Spacing.kt`,
    filter: 'isSpacing',
    formatter: 'android/spacings',
    packageName: theme.spacings.packageName,
  },
  colorsTemplate: {
    source: join(__dirname, './templates/colors.ejs'),
    destination: `${theme.colors}/Color.kt`,
    filter: 'color',
    formatter: 'android/compose/colors',
    packageName: theme.colors.packageName,
  },
  headerTemplate: join(__dirname, './templates/header.ejs'),
})
