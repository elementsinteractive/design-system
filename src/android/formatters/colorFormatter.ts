import Color from 'tinycolor2'
import { useTemplate } from '../../utils'
import type { Formatter } from 'style-dictionary'
import { FormatterConfig } from './types'

export const colorFormatter: (config: FormatterConfig) => Formatter =
  (config: FormatterConfig) =>
  ({ dictionary }) => {
    const { template, header, packageName } = config

    const tokens = dictionary.allTokens

    const toArgb = color => {
      const colorStr = Color(color).toHex8()

      return colorStr.slice(6) + colorStr.slice(0, 6)
    }

    const colors = tokens.map(({ name, value }) => ({
      name,
      value: toArgb(value),
    }))

    return useTemplate(template)({
      header,
      colors,
      packageName,
    })
  }
