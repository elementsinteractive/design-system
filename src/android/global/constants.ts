import { join } from 'path'

import { AndroidGlobalOptions } from '../types'

export const ColorTypeFilename = 'ColorTokens.kt'

export const SpacingTypeFilename = 'SpacingTokens.kt'

export const GLOBAL_TEMPLATES = (options: AndroidGlobalOptions) => ({
  colorsTypeTemplate: createColorsTypeTemplate(options),
  spacingsTypeTemplate: createSpacingsTypeTemplate(options),
  headerTemplate: join(__dirname, './templates/header.ejs'),
})

const createSpacingsTypeTemplate = (options: AndroidGlobalOptions) => ({
  source: join(__dirname, './templates/spacingTokens.ejs'),
  destination: `${options.destination}/${SpacingTypeFilename}`,
  filter: 'spacingsType',
  formatter: 'android/spacingsType',
  packageName: options.packageName,
})

const createColorsTypeTemplate = (options: AndroidGlobalOptions) => ({
  source: join(__dirname, './templates/colorTokens.ejs'),
  destination: `${options.destination}/${ColorTypeFilename}`,
  filter: 'colorsType',
  formatter: 'android/compose/colorsType',
  packageName: options.packageName,
})
