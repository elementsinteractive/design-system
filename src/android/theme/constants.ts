import { join } from 'path'
import { AndroidThemeOptions } from '../types'

export const PaletteFilename = 'Palette.kt'
export const SpacingFilename = 'Spacing.kt'
export const TypeFilename = 'Type.kt'
export const ColorTokensFilename = 'ColorTokens.kt'

export const THEME_TEMPLATES = (options: AndroidThemeOptions) => ({
  typographyTemplate: createTypographyTemplate(options),
  spacingsTemplate: createSpacingsTemplate(options),
  paletteTemplate: createPaletteTemplate(options),
  colorTokensTemplate: createColorTokensTemplate(options),
  headerTemplate: join(__dirname, './templates/header.ejs'),
})

const createTypographyTemplate = ({ theme }: AndroidThemeOptions) => ({
  source: join(__dirname, './templates/typographies.ejs'),
  destination: `${theme.typography.destination}/${TypeFilename}`,
  filter: 'typography',
  formatter: 'android/compose/typography',
  packageName: theme.typography.packageName,
})

const createSpacingsTemplate = (options: AndroidThemeOptions) => ({
  source: join(__dirname, './templates/base/spacings.ejs'),
  destination: `${options.theme.spacings.destination}/${SpacingFilename}`,
  filter: 'spacings',
  formatter: 'android/spacings',
  packageName: options.theme.spacings.packageName,
  sharedPackageName: options.globalPackageName,
})

const createPaletteTemplate = ({ theme }: AndroidThemeOptions) => ({
  source: join(__dirname, './templates/shared/palette.ejs'),
  destination: `${theme.palette.destination}/${PaletteFilename}`,
  filter: 'color',
  formatter: 'android/compose/palette',
  packageName: theme.palette.packageName,
  themeName: theme.name,
})

const createColorTokensTemplate = ({ theme }: AndroidThemeOptions) => ({
  source: join(__dirname, './templates/shared/colorTokens.ejs'),
  destination: `${theme.colors.destination}/${ColorTokensFilename}`,
  filter: 'color',
  formatter: 'android/compose/colorTokens',
  packageName: theme.colors.packageName,
  themeName: theme.name,
})
