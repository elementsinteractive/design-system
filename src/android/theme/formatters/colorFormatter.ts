import Color from 'tinycolor2'
import { Formatter } from 'style-dictionary'

import { FormatterConfig } from '../../types'
import { useTemplate } from '../../../utils'

export const colorFormatter =
  (
    config: FormatterConfig,
    globalPackageName: string,
    themeName: string,
  ): Formatter =>
  ({ dictionary }) => {
    const tokens = dictionary.allTokens
    const { template, header, packageName } = config

    const colors = tokens.map(({ name, value }) => ({
      colorName: name,
    }))

    return useTemplate(template)({
      header,
      colors,
      packageName,
      globalPackageName,
      themeName,
    })
  }
