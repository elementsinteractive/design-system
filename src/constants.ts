export const LOG_PREFIX = '[Design]'