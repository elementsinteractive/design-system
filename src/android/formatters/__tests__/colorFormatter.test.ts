import createDictionary from 'style-dictionary/lib/utils/createDictionary'
import createFormatArgs from 'style-dictionary/lib/utils/createFormatArgs'

import { TEMPLATES } from '../../constants'
import { colorFormatter } from '../colorFormatter'
import { FormatterConfig } from '../../types'
import { HEADER } from '../../../common/formatters/__tests__/constants'
import properties from '../../../common/formatters/__tests__/colorProperties'

describe('colorFormatter', () => {
  it('can handle dictionary with color tokens', () => {
    const { colorsTemplate } = TEMPLATES('')

    const colorConfig: FormatterConfig = {
      template: colorsTemplate.source,
      header: HEADER,
      packageName: 'nl.elements.skeleton',
    }

    const dictionary = createDictionary({ properties })
    const formatterArgs = createFormatArgs({ dictionary, platform: {} })

    const formatter = colorFormatter(colorConfig)(formatterArgs)

    expect(formatter).toMatchSnapshot()
  })
})
