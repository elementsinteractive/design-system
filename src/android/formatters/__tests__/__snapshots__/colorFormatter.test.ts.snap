// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`colorFormatter can handle dictionary with color tokens 1`] = `
"package nl.elements.skeleton

/* Test header */
import androidx.compose.ui.graphics.Color
import com.airbnb.android.showkase.annotation.ShowkaseColor

@ShowkaseColor(name = \\"primary\\", group = \\"material\\")
val primary = Color(0xffffffff)

@ShowkaseColor(name = \\"background\\", group = \\"material\\")
val background = Color(0xfff4f4f4)

@ShowkaseColor(name = \\"primaryVariant\\", group = \\"material\\")
val primaryVariant = Color(0xff000000)

@ShowkaseColor(name = \\"secondary\\", group = \\"material\\")
val secondary = Color(0xffff495b)

"
`;
